cartographer_grpc/framework/rpc.cc

```diff
@@ -305,16 +326,23 @@ ActiveRpcs::~ActiveRpcs() {
 
 Rpc* ActiveRpcs::Add(std::unique_ptr<Rpc> rpc) {
   cartographer::common::MutexLocker locker(&lock_);
-  const auto result = rpcs_.emplace(rpc.release());
+  std::shared_ptr<Rpc> shared_rpc = std::move(rpc);
+  const auto result = rpcs_.emplace(shared_rpc.get(), shared_rpc);
   CHECK(result.second) << "RPC already active.";
-  return *result.first;
+  return shared_rpc.get();
+}
+
+std::shared_ptr<Rpc> ActiveRpcs::Get(Rpc* rpc) EXCLUDES(lock_) {
+  cartographer::common::MutexLocker locker(&lock_);
+  auto it = rpcs_.find(rpc);
+  CHECK(it != rpcs_.end()) << "RPC " << rpc << " not found.";
+  return it->second;
 }
 
 bool ActiveRpcs::Remove(Rpc* rpc) {
   cartographer::common::MutexLocker locker(&lock_);
   auto it = rpcs_.find(rpc);
   if (it != rpcs_.end()) {
-    delete rpc;
```

Comment:
Who deletes Rpc now?

Reply:
The `shared_ptr` does, when the last one goes out of scope. The one stored in ActiveRpcs is deleted in the line below this one.